File: packages/training/seed.py
"""Deterministic seed -> latent, bit-identical to packages/core/src/seed.ts."""

import numbers
from collections.abc import Iterable

import numpy as np

LATENT_DIM = 32
SEED_VERSION = 1
_MASK = 0xFFFFFFFF
_GOLDEN = 0x9E3779B9
_DRAWS_PER_DIM = 12
_TWO_32 = 4294967296.0


def to_uint32_seed(seed: int) -> int:
    if isinstance(seed, bool) or not isinstance(seed, numbers.Integral):
        raise TypeError(f"seed must be an int, got {seed!r}")
    seed = int(seed)
    if abs(seed) > 2**53 - 1:
        raise TypeError("seed must be a JavaScript safe integer")
    return seed & _MASK


def seed_to_latent(seed: int) -> np.ndarray:
    x = to_uint32_seed(seed)
    if x == 0:
        x = _GOLDEN
    z = np.empty(LATENT_DIM, dtype=np.float32)
    for i in range(LATENT_DIM):
        acc = 0.0
        for _ in range(_DRAWS_PER_DIM):
            x ^= (x << 13) & _MASK
            x ^= x >> 17
            x ^= (x << 5) & _MASK
            acc += x / _TWO_32
        z[i] = np.float32(acc - 6.0)
    return z


def latents_for_seeds(seeds: Iterable[int]) -> np.ndarray:
    return np.stack([seed_to_latent(s) for s in seeds]).astype(np.float32)

File: packages/core/src/seed.ts
/** Latent dimension of the shipped decoder. */
export const LATENT_DIM = 32;
/** Version of the seed → latent algorithm. Bump if the algorithm changes. */
export const SEED_VERSION = 1;

const GOLDEN = 0x9e3779b9;
const DRAWS_PER_DIM = 12;
const TWO_32 = 4294967296;

/** Reduce any safe integer to an unsigned 32-bit seed (two's complement for negatives). */
export function toUint32Seed(seed: number): number {
  if (!Number.isSafeInteger(seed)) {
    throw new TypeError(`seed must be a safe integer, got ${String(seed)}`);
  }
  return seed >>> 0;
}

/**
 * Deterministic seed → latent. xorshift32 drives an Irwin-Hall approximation
 * of a standard normal (sum of 12 uniforms minus 6). Uses only integer ops,
 * additions and one division so Python and TypeScript agree bit for bit.
 */
export function seedToLatent(seed: number): Float32Array {
  let x = toUint32Seed(seed);
  if (x === 0) x = GOLDEN;
  const z = new Float32Array(LATENT_DIM);
  for (let i = 0; i < LATENT_DIM; i++) {
    let acc = 0;
    for (let k = 0; k < DRAWS_PER_DIM; k++) {
      x ^= x << 13;
      x >>>= 0;
      x ^= x >>> 17;
      x >>>= 0;
      x ^= x << 5;
      x >>>= 0;
      acc += x / TWO_32;
    }
    z[i] = acc - 6;
  }
  return z;
}
